network/p2p/network.go

```diff
@@ -132,21 +172,23 @@ func (n *Network) Register(channel network.Channel, engine network.Engine) (netw
 		Str("channel_id", channel.String()).
 		Msg("channel successfully registered")
 
-	// create a cancellable child context
-	ctx, cancel := context.WithCancel(n.ctx)
+	respChan := make(chan *Conduit)
 
-	// create the conduit
-	conduit := &Conduit{
-		ctx:       ctx,
-		cancel:    cancel,
-		channel:   channel,
-		publish:   n.publish,
-		unicast:   n.unicast,
-		multicast: n.multicast,
-		close:     n.unregister,
+	select {
```

Comment:
can you add some comments describing the interactions between register and the worker function added in Start. It's not obvious what's going on to someone that's unfamiliar with the code what.

Is that worker intended to stay alive forever? Is register something that's called many times?

Reply:
The difference is that the process which responds to register requests uses the context passed to Start(), meaning it will terminate when the context is cancelled.

This is the desired behavior, because Register() calls made after the Network has been cancelled should fail. Previously this same behavior was implemented by storing a context directly inside the Network struct, which is then canceled when Done() is called. However, storing contexts inside structs is an anti-pattern: https://go.dev/blog/context-and-structs. Furthermore, going forward the Done() method will no longer be used to trigger shutdown of a component, but merely to check when shutdown has completed.

Now, before anyone else has the chance to point it out, I admit that exposing the `ShutdownSignal` API from `ComponentManager` is somewhat of a loop hole to the contexts-in-structs "rule". I'm hoping that eventually it may
be deprecated, but for now having it there may make the transition to the `Component` interface easier. Also, wrapping the `ShutdownSignal` in a `ComponentManager` hopefully makes it more clear what the scope of the shutdown signal applies to, whereas storing a context is more ambiguous because there's no way to know whether the context affects routines outside the scope of the component itself.